src/keywrap/keyprovider.rs

```diff
@@ -0,0 +1,572 @@
+            };
+        } else if self.attrs.grpc.as_ref().is_some() {
+            let rt = Runtime::new().unwrap().block_on(get_provider_grpc_output(serialized_input, self.attrs.grpc.as_ref().unwrap(), OP_KEY_UNWRAP.to_string()));
+            protocol_output = match rt {
+                Ok(v) => v,
+                Err(e) => return Err(anyhow!("Error while key provider {:?} operation, from grpc provider error, error: {:?}", OP_KEY_UNWRAP.to_string(), e))
+            };
+        }
+        let key_unwrap_results = match protocol_output.key_unwrap_results {
+            Some(x) => x,
+            None => return Err(anyhow!("keyprovider: {:?} on {:?} operation results are empty",self.provider,OP_KEY_UNWRAP.to_string()))
+        };
+
+        Ok(key_unwrap_results.opts_data)
+    }
+
+
+    fn annotation_id(&self) -> String {
+        format!("{}{}", "org.opencontainers.image.enc.keys.provider.", self.provider)
+    }
+
+    fn no_possible_keys(&self, _dc_param: &HashMap<String, Vec<Vec<u8>>, RandomState>) -> bool {
+        unimplemented!()
+    }
+}
+
+async fn get_provider_grpc_output(input: Vec<u8>, conn: &str, operation: String) -> Result<KeyProviderKeyWrapProtocolOutput> {
+    let mut protocol_output = KeyProviderKeyWrapProtocolOutput { key_wrap_results: None, key_unwrap_results: None };
+    let uri = conn.parse::<Uri>().unwrap();
+    // create a channel ie connection to server
+    let channel = match tonic::transport::Channel::builder(uri)
+        .connect()
+        .await {
+        Ok(x) => x,
+        Err(e) => return Err(anyhow!("Error while creating channel: {:?}",e))
+    };
+    let mut client = keyproviderpb::key_provider_service_client::KeyProviderServiceClient::new(channel);
+    let request = tonic::Request::new(keyproviderpb::KeyProviderKeyWrapProtocolInput { key_provider_key_wrap_protocol_input: input });
+    if operation == OP_KEY_WRAP {
+        let grpc_output = match client.wrap_key(request).await {
+            Ok(resp) => resp,
+            Err(_) => return Err(anyhow!("Error from grpc request method on {:?} operation", OP_KEY_WRAP.to_string()))
+        };
+        protocol_output = match serde_json::from_slice(&grpc_output.into_inner().key_provider_key_wrap_protocol_output) {
+            Ok(x) => x,
+            Err(_) => return Err(anyhow!("Error while deserializing grpc output on {:?} operation", OP_KEY_WRAP.to_string()))
+        }
+    } else if operation == OP_KEY_UNWRAP {
+        let grpc_output = match client.un_wrap_key(request).await {
+            Ok(resp) => resp,
+            Err(_) => return Err(anyhow!("Error from grpc request method on {:?} operation", OP_KEY_UNWRAP.to_string()))
+        };
+        protocol_output = match serde_json::from_slice(&grpc_output.into_inner().key_provider_key_wrap_protocol_output) {
+            Ok(x) => x,
+            Err(_) => return Err(anyhow!("Error while deserializing grpc output on {:?} operation", OP_KEY_UNWRAP.to_string()))
+        };
+    }
```

Comment:
`map_err` for more readable code.

Reply:
I want specifically return error here, map_err returns a result, again it has to  check with is_ok condition and process further, that will lead to too many if else statements. I don't think we can return error directly from

Reply:
@pravinrajr9 `let protocol_output = serde_json::from_slice().map_err(|_| anyhow!())?;` will return a `Result::Err` if `from_slice` fails.

Reply:
super cool thanks